---
title: "Figure 2"
author: "Florian WÃ¼nnemann"
output: html_document
---

```{r setup, dev='svg'}
library(tidyverse)
library(data.table)
library(cowplot)
library(ggridges)
library(ggrepel)
library(GenomicRanges)
library(biomaRt)
library(patchwork)
library(here)

source("./Figure_functions.R")

theme_set(theme_cowplot_custom)
```

# Figure 2

## Figure 2A : UMAP plot

The code to calculate PCA and UMAP embeddings can be found in [figure1.calculate_umap.Rmd](./figure1.calculate_umap.Rmd).

```{r}
## Read in UMAP mappings
umap_mappings <- fread(here("./data/figure2.UMAP.final_mappings.tsv"))

## Group by Cas variant for ellipse
umap_mappings <- umap_mappings %>%
  group_by(Cas_variant)

## Replace simple Cas flavour names by full names
umap_mappings$Cas_variant <- gsub("KRAB","CRISPRi",umap_mappings$Cas_variant)
umap_mappings$Cas_variant <- gsub("VP64","CRISPRa",umap_mappings$Cas_variant)

figure_2a <- ggplot(umap_mappings,aes(UMAP1,UMAP2)) +
  geom_point(aes(fill = Cas_variant),size = 3,colour = "black", pch = 21, alpha = 1) +
  xlim(min(umap_mappings$UMAP1)-1,max(umap_mappings$UMAP1)+1) +
  ylim(min(umap_mappings$UMAP2)-1,max(umap_mappings$UMAP2)+1) +
  # geom_mark_ellipse(aes(fill = Cas_variant , label = Cas_variant),
  #                   alpha = 0.4) +
  # scale_colour_manual("Cas flavour",
  #                     values = c("#EFC7B1","#F595A2","#C9CCA8"),
  #                     labels = c("Cas9","dCas9-KRAB","dCas9-VP64")) +
  scale_fill_manual("Cas9 variant",
                      values = c("#EFC7B1","#F595A2","#C9CCA8"),
                      labels = c("Cas9","CRISPRi","CRISPRa")) +
  theme(legend.position = "bottom",
        legend.title = element_text(colour="black", 
                                      face="bold"),
        # axis.line = element_blank(),
        # axis.ticks = element_blank(),
        # axis.text =  element_blank(),
        axis.title = element_text(colour="black", 
                                      face="bold")) +
  labs(x = "UMAP 1",
       y = "UMAP 2") 
  

figure_2a
```

## Figure 2B


```{r}
## Load data
mageck_all_pooled_results <- fread(here("./data/figure2.global_results.all_pooled.with_plasmid.tsv"))

## Read in counts
raw_counts <- fread(here("./data/figure2.all-sample_counts.annotated.tsv"))
```


```{r}
## reformat MAGECK table

## replace special strings in column names
colnames(mageck_all_pooled_results) <- gsub("\\|","_",colnames(mageck_all_pooled_results))
colnames(mageck_all_pooled_results) <- gsub("\\-","_",colnames(mageck_all_pooled_results))

mageck_all_pooled_results <- mageck_all_pooled_results %>%
  mutate("significant" = if_else(FACS_samples_fdr <= 0.1,"yes","no"))

colnames(mageck_all_pooled_results) <- gsub("Gene","Locus",colnames(mageck_all_pooled_results))

## Add sgRNA groups
count_table_annotated_sub <- raw_counts %>%
  dplyr::select(Locus,sgRNA_group)

mageck_all_pooled_results_anno <- left_join(mageck_all_pooled_results,count_table_annotated_sub,by=c("Locus"))

mageck_all_pooled_results_anno <- mageck_all_pooled_results_anno %>%
  mutate("sgRNA_group" = if_else(grepl("Ess_",Locus),"essential_genes",sgRNA_group)) %>%
  unique()

## add different groupings for sgRNA
mageck_all_pooled_results_anno <- mageck_all_pooled_results_anno %>%
  mutate("sgrna_supergroup" = if_else(sgRNA_group == "essential_genes","essential_genes","Rest"))
```


```{r}
## replace some columns for plotting
mageck_all_pooled_results_anno$sgrna_supergroup <- gsub("essential_genes","Essential genes",mageck_all_pooled_results_anno$sgrna_supergroup)
mageck_all_pooled_results_anno$sgrna_supergroup <- gsub("Rest","Other sgRNA",mageck_all_pooled_results_anno$sgrna_supergroup)
mageck_all_pooled_results_anno$Cas_variant <- gsub("CAS9","Cas9",mageck_all_pooled_results_anno$Cas_variant)

## Add Label for geom repel
mageck_all_pooled_results_anno <- mageck_all_pooled_results_anno %>%
  mutate("Label" = if_else(FACS_samples_fdr < 0.1 & sgRNA_group == "CAD_loci",Locus,""))
```

### Calculate statistics

```{r}
## Cas9
cas9_essential <- subset(mageck_all_pooled_results_anno,Cas_variant == "Cas9" & sgrna_supergroup == "Essential genes")
cas9_other <- subset(mageck_all_pooled_results_anno,Cas_variant == "Cas9" & sgrna_supergroup == "Other sgRNA")
cas9_results <- ks.test(cas9_essential$FACS_samples_beta,cas9_other$FACS_samples_beta, alternative = "two.sided")
cas9_results$p.value

## dCas9-KRAB
krab_essential <- subset(mageck_all_pooled_results_anno,Cas_variant == "KRAB" & sgrna_supergroup == "Essential genes")
krab_other <- subset(mageck_all_pooled_results_anno,Cas_variant == "KRAB" & sgrna_supergroup == "Other sgRNA")
krab_results <- ks.test(krab_essential$FACS_samples_beta,krab_other$FACS_samples_beta, alternative = "two.sided")
krab_results$p.value

## dCas9-VP64
vp64_essential <- subset(mageck_all_pooled_results_anno,Cas_variant == "VP64" & sgrna_supergroup == "Essential genes")
vp64_other <- subset(mageck_all_pooled_results_anno,Cas_variant == "VP64" & sgrna_supergroup == "Other sgRNA")
vp64_results <- ks.test(vp64_essential$FACS_samples_beta,vp64_other$FACS_samples_beta, alternative = "two.sided")
vp64_results$p.value
```


### Plot Figure

```{r, warnings = FALSE}
## Plot results
mageck_all_pooled_results_anno <- mageck_all_pooled_results_anno %>%
  mutate("plot_CAS_variant" = if_else(Cas_variant == "KRAB","CRISPRi",
                                      if_else(Cas_variant == "VP64","CRISPRa","Cas9")))

mageck_all_pooled_results_anno$plot_CAS_variant <- factor(mageck_all_pooled_results_anno$plot_CAS_variant,
                                                          levels = c("Cas9","CRISPRi","CRISPRa"))

figure_2b <- ggplot(mageck_all_pooled_results_anno,aes(FACS_samples_beta,
                                                       sgrna_supergroup,
                                                       fill = sgrna_supergroup,
                                                       label = Label)) +
  geom_density_ridges(color = "black") +
  theme_cowplot() +
  facet_grid(plot_CAS_variant~.) +
  scale_fill_brewer("sgRNA type",
    palette = "Dark2") +
  labs(x = "Beta (sorted cells vs plasmid)",
       y = "Density") +
  geom_vline(xintercept = 0, linetype = 2, size = 1) +
  theme(axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.line.y = element_blank(),
        legend.position = "bottom",
        legend.title = element_text(colour="black", 
                                      face="bold"),
        strip.background=element_rect(fill="grey20"),
        strip.text.y = element_text(colour = 'white',
                                  face = "bold",
                                  angle = 0)
  ) 
  geom_text_repel()
  
figure_2b
```

## Figure 2 C

# Load final MAGECK results

```{r}
final_mageck <- fread(file = here("./data/figure2.final_locus_results.tsv"))

## Exclude loci with only 1 sgRNA
final_mageck <- final_mageck %>%
  subset(sgRNA > 1)
```

# Positive controls for adhesion molecules

```{r}
## Add specific positive control as group
final_mageck_mod <- final_mageck %>%
  mutate("sgrna_supergroup" = if_else(FACS_protein == "SELE" & grepl("SELE|Sele",Locus),"positive_control",
                                      if_else(FACS_protein == "CD31" & grepl("PECAM|Pecam",Locus),"positive_control",
                                              if_else(FACS_protein == "ICAM" & grepl("ICAM|Icam",Locus),"positive_control",
                                                      if_else(FACS_protein == "VCAM" & grepl("VCAM|Vcam",Locus),"positive_control",sgRNA_group))))
  ) 

## add different groupings
final_mageck_mod <- final_mageck_mod %>%
  mutate("sgrna_supergroup" = if_else(grepl("Ess_",Locus),"essential_genes",sgrna_supergroup))

## Rename columns for plotting
final_mageck_mod$sgrna_supergroup <- gsub("CAD_loci","CAD loci",final_mageck_mod$sgrna_supergroup)
final_mageck_mod$sgrna_supergroup <- gsub("known_genes","Known genes",final_mageck_mod$sgrna_supergroup)
final_mageck_mod$sgrna_supergroup <- gsub("positive_promotors","Promotors",final_mageck_mod$sgrna_supergroup)

## Rename Cas proteins
final_mageck_mod <- final_mageck_mod %>%
  mutate("Cas_variant" = if_else(Cas_variant == "KRAB","CRISPRi",
                                 if_else(Cas_variant == "VP64","CRISPRa",Cas_variant))) %>%
  mutate("FACS_protein" = if_else(FACS_protein =="VCAM","VCAM1",
                                  if_else(FACS_protein == "ICAM","ICAM1",FACS_protein)))

final_mageck_mod$FACS_protein <- factor(final_mageck_mod$FACS_protein,
                                        levels = c("SELE","VCAM1","ICAM1","ROS","NO","Ca"))
```

## Plot function for ranked plot

```{r}
plot_ranked_beta <- function(rank_data_frame,
                             color_column,
                             label_column){
  
  colors <- c("Exon" = "darkgreen", "Promoter" = "darkblue", "Other sgRNA" = "darkgrey")
  
  ranked_beta_pos_ctr_plot <- ggplot(rank_data_frame,aes(rank,Beta, 
                                                         color = get(color_column),
                                                         label = get(label_column)
                                                         )) +
  geom_point(data = subset(rank_data_frame,get(color_column) == "Other sgRNA"),
             size = 1.5, color = "darkgrey") +
  geom_point(data = subset(rank_data_frame,get(color_column) == "Exon"),
             size = 3, color = "darkgreen", alpha = 0.6) +
  geom_point(data = subset(rank_data_frame,get(color_column) == "Promoter"),
             size = 3, color = "darkblue" , alpha = 0.6) +
  facet_grid(Cas_variant ~ FACS_protein) +
  scale_color_manual("sgRNA target",
                     values = colors) +
  theme_cowplot() +
  labs(x = "Rank",
       y = "Beta") +
  theme(
  axis.title.x = element_text(face="bold"),
  axis.title.y = element_text(face="bold"),
  legend.position = "none",
  legend.title = element_text(colour="black", 
                                      face="bold"),
        strip.background=element_rect(fill="grey20"),
        strip.text = element_text(colour = 'white',
                                  face = "bold")
  ) +
  geom_hline(yintercept = 0, linetype = 2) +
  scale_y_continuous(expand = expansion(mult = 0.15)) +
  scale_x_continuous(breaks = c(1,500,1000,1500,2000),
                     expand = expansion(mult = 0.15)) 
  
  return(ranked_beta_pos_ctr_plot)
}

```

## Cas9 figure

```{r}
## general plotting settings
repel_label_size <- 2
```


```{r}
cas9_positive_controls <- final_mageck_mod %>%
  subset(FACS_protein %in% c("SELE","VCAM1","ICAM1")) %>%
  subset(Cas_variant == "Cas9")

cas9_positive_controls <- cas9_positive_controls %>%
  group_by(FACS_protein) %>%
  mutate("rank" = rank(Beta, ties.method = "first")) %>%
  mutate("significant" = if_else(FDR <= 0.1,"yes","no")) %>%
  mutate("highlight_Locus" = if_else(sgrna_supergroup == "positive_control" & significant == "yes",Locus,"")) %>%
  mutate("highlight_Locus_group" = if_else(sgrna_supergroup == "positive_control",
                                           if_else(grepl("Pos_",Locus),"Exon",
                                                   "Promoter"),"Other sgRNA")) %>%
    mutate("highlight_Locus_group_label" = if_else(sgrna_supergroup == "positive_control",
                                                   paste(rank,Locus,sep=":"),"")) %>%
  ungroup()%>%
  arrange(rank)

  ## Set order of color groups
  cas9_positive_controls$highlight_Locus_group <- factor(cas9_positive_controls$highlight_Locus_group,
                                                       levels = c("Exon","Promoter","Other sgRNA"))
  
  test <- subset(cas9_positive_controls,highlight_Locus_group_label != "")
  test <- as.data.frame(test)
  
  ## Replace SELE with E-Selectin
  cas9_positive_controls <- cas9_positive_controls %>% 
    mutate("FACS_protein" = gsub("SELE","E-Selectin",FACS_protein))

  cas9_positive_controls_plot <- plot_ranked_beta(cas9_positive_controls,
                                                  color_column = "highlight_Locus_group",
                                                  label_column = "highlight_Locus_group_label") +
  geom_label_repel(size = repel_label_size,
                   max.overlaps = 2000,
                   force_pull   = 0) +
  labs(x = "")
  
  cas9_positive_controls_plot
```

## KRAB

```{r}
krab_positive_controls <- final_mageck_mod %>%
  subset(FACS_protein %in% c("SELE","VCAM1","ICAM1")) %>%
  subset(Cas_variant == "CRISPRi")

krab_positive_controls <- krab_positive_controls %>%
  group_by(FACS_protein) %>%
  mutate("rank" = rank(Beta, ties.method = "first")) %>%
  mutate("significant" = if_else(FDR <= 0.1,"yes","no")) %>%
  mutate("highlight_Locus" = if_else(sgrna_supergroup == "positive_control" & significant == "yes",Locus,"")) %>%
  mutate("highlight_Locus_group" = if_else(sgrna_supergroup == "positive_control",
                                           if_else(grepl("Pos_",Locus),"Exon",
                                                   "Promoter"),"Other sgRNA")) %>%
    mutate("highlight_Locus_group_label" = if_else(sgrna_supergroup == "positive_control",
                                                   paste(rank,Locus,sep=":"),"")) %>%
  ungroup()%>%
  arrange(rank)

  ## Set order of color groups
  krab_positive_controls$highlight_Locus_group <- factor(krab_positive_controls$highlight_Locus_group,
                                                       levels = c("Exon","Promoter","Other sgRNA"))
  
  krab_positive_controls$FACS_protein <- factor(krab_positive_controls$FACS_protein,
                                                levels = c("SELE","ICAM1","VCAM1"))

  krab_positive_controls_plot <- plot_ranked_beta(krab_positive_controls,
                                                  color_column = "highlight_Locus_group",
                                                  label_column = "highlight_Locus_group_label")
  
  krab_positive_controls_plot <- krab_positive_controls_plot + 
    theme(strip.background.x = element_blank(),
          strip.text.x = element_blank()) +
    # geom_label_repel(force = 2,
    #                   xlim = c(50,2000),
    #                   ylim = c(-4,0),
    #                   point.padding = 0.1,
    #                   size = repel_label_size,
    #                   #label.padding = 0.01,
    #                 max.overlaps = 1000) +
  geom_label_repel(size = repel_label_size,
                   max.overlaps = 2000,
                   force_pull   = 0) +
    labs(x = "")
  
  krab_positive_controls_plot
```


## CRISPRa 

```{r}
vp64_positive_controls <- final_mageck_mod %>%
  subset(FACS_protein %in% c("SELE","VCAM1","ICAM1")) %>%
  subset(Cas_variant == "CRISPRa")

vp64_positive_controls <- vp64_positive_controls %>%
  group_by(FACS_protein) %>%
  mutate("rank" = rank(dplyr::desc(Beta), ties.method = "first")) %>%
  mutate("significant" = if_else(FDR <= 0.1,"yes","no")) %>%
  mutate("highlight_Locus" = if_else(sgrna_supergroup == "positive_control" & significant == "yes",Locus,"")) %>%
  mutate("highlight_Locus_group" = if_else(sgrna_supergroup == "positive_control",
                                           if_else(grepl("Pos_",Locus),"Exon",
                                                   "Promoter"),"Other sgRNA")) %>%
    mutate("highlight_Locus_group_label" = if_else(sgrna_supergroup == "positive_control",
                                                   paste(rank,Locus,sep=":"),"")) %>%
  ungroup()%>%
  arrange(rank)

  ## Set order of color groups
  vp64_positive_controls$highlight_Locus_group <- factor(vp64_positive_controls$highlight_Locus_group,
                                                       levels = c("Exon","Promoter","Other sgRNA"))
  
  vp64_positive_controls$FACS_protein <- factor(vp64_positive_controls$FACS_protein,
                                                levels = c("SELE","ICAM1","VCAM1"))

  vp64_positive_controls_plot <- plot_ranked_beta(vp64_positive_controls,
                                                  color_column = "highlight_Locus_group",
                                                  label_column = "highlight_Locus_group_label")
  
  vp64_positive_controls_plot <- vp64_positive_controls_plot  + 
    theme(strip.background.x = element_blank(),
        strip.text.x = element_blank()) +
  geom_label_repel(size = repel_label_size,
                   max.overlaps = 2000,
                   force_pull   = 0)
  
vp64_positive_controls_plot

```


## Assemble final figure

```{r}
top_figures <- (figure_2a |  figure_2b) +   
  plot_layout(ncol = 2,
              widths = c(1,1))

pos_control_figures <- (cas9_positive_controls_plot / krab_positive_controls_plot/ vp64_positive_controls_plot) +   
  plot_layout(nrow = 3,
              heights = c(1,1,1)) 

final_figure <- top_figures / pos_control_figures +
  plot_layout(nrow = 2,
              heights = c(1,2.5)) +
  plot_annotation(tag_levels = 'A') & 
  theme(plot.tag = element_text(size = 20))

plot_height <- 10
plot_width <- 11

save_plot(final_figure,
          filename = here("./figures/Figure_2.eps"),
          base_height = plot_height,
          base_width = plot_width)
```
